Fix event generics.

Cybermaxke opened the pull request:
**Common** | [Forge](https://github.com/SpongePowered/SpongeForge/pull/1929)

Fixes generic events that use complex generic types, like `DataRegistration<?,?>`, `Key<? extends BaseValue<?>>`, etc.
Also fixes a issue that caused the cause stack manager to be used on the client when it's not supposed to.

@gabizou

Cybermaxke committed:
Improved the generic event.

This fixes generic events that have complex generic types, like DataRegistration<?,?>, Key<?>, etc. Also fixes the MyHomes plugin for testing (make it run without errors). Remove event post method on client thread, was already fixed before.

Cybermaxke committed:
Init the console permissions on the client once the server is available.

randombyte-developer commented:
:+1: I didn't bother debugging the non-working listener and kept the 'old' PreInit event for the registration of Keys. Thanks for fixing it!

gabizou commented on testplugins/src/main/java/org/spongepowered/test/myhomes/MyHomes.java:
```diff
@@ -67,61 +68,67 @@
             .name("Default Home")
             .query(DataQuery.of("DefaultHome"))
             .build();
+        event.register(DEFAULT_HOME);
```
Unless you changed the `SpongeKeyBuilder` implementation, this is redundant and should throw an exception for attempting double registration.

gabizou commented on src/main/java/org/spongepowered/common/SpongeImpl.java:
```diff
@@ -197,18 +197,10 @@ public static boolean postEvent(Event event) {
     public static void postState(GameState state, GameStateEvent event) {
         check(game);
         game.setState(state);
-        ((SpongeEventManager) game.getEventManager()).post(event, true);
```
The reason why this was added is because on clients, this will not work unless allowing a client thread event is true.

Cybermaxke committed:
Keys must be registered through the event or registry.

- Added support for non static inner classes in the typetoken helper
- The plugin id for keys will be retrieved from the current plugin container (old method is deprecated, but still working)

Cybermaxke committed:
Improved the TypeTokenHelper#isAssignable method.

gabizou commented:
I'm curious why you are requiring `Key`s to be registered through the event instead of through the `build()` method as it was originally designed (much like `DataRegistration` for that matter).

Cybermaxke commented:
@gabizou I think we should be consistent how we register stuff. And if we aren't going to use the `register` method then there is no point to register things inside event, might as well do it in the init event. `DataRegistration` would require it's `buildAndRegister` to be changed into `build` to fit into this concept.

Cybermaxke committed:
Fix that the causestackmanager was used on the client.

Cybermaxke commented:
@gabizou This PR should also be merged into bleeding.

dualspiral commented:
@Cybermaxke I've pulled all the latest `stable-7` changes into `bleeding` now.
